### PR TITLE
support certificate expiry status

### PR DESCRIPTION
**Pull request by suhasHere**

### CODE REVIEW

File: lib/hpke/src/certificate.cpp [hunk 1]
```diff
@@ -234,6 +234,34 @@ struct Certificate::ParsedCertificate
     return make_typed_unique<EVP_PKEY>(X509_get_pubkey(x509.get()));
   }
 
+  Certificate::Status status() const
+  {
+    auto* notBefore = X509_get_notBefore(x509.get());
+    auto ret = X509_cmp_current_time(notBefore);
+    if (ret == 0) {
```
Comment (dimarusyy) [inline lib/hpke/src/certificate.cpp:241]: I'm not sure but `switch(X509_cmp_current_time(notBefore)) {...}` could potentially give better branch prediction.
Btw, did does it make sense to add attributes `[[likely]]`?

**Review by dimarusyy — APPROVED** [review]

File: lib/hpke/test/certificate.cpp [hunk 2]
```diff
@@ -470,3 +472,64 @@ TEST_CASE("Test Subject Parsing")
...
+    "0405130831312d32322d3333302a300506032b6570032100cb6d233470f884eaa6e5a3b958"
+    "e7e68eff3fee146432ea526128171a33f8a403a34e304c300e0603551d0f0101ff04040302"
+    "02a4300f0603551d130101ff040530030101ff300d0603551d0e0406040459e6b3dd301a06"
+    "03551d1104133011810f7573657240646f6d61696e2e636f6d300506032b6570034100308a"
+    "a37cfbcd06e19b7a0728c5c970b38df5eb93d478970868ce6398a6a963b2c570edfd9dc62f"
+    "4d134de11eca367f9d967d6eae14192454770a2fc278963602");
+
+  auto root = Certificate{ root_der };
+  auto status = root.status();
+  REQUIRE(status == Certificate::Status::inactive);
```
Comment (bifurcation) [inline lib/hpke/test/certificate.cpp:493]:
```suggestion
  REQUIRE(root.status() == Certificate::Status::inactive);
```

File: lib/hpke/test/certificate.cpp [hunk 3]
```diff
@@ -470,3 +472,64 @@ TEST_CASE("Test Subject Parsing")
...
+    "e4e7d2b0606050b2e0edcfc8d6390b373e21f08116910b");
+
+  auto root = Certificate{ root_der };
+  auto status = root.status();
+  REQUIRE(status == Certificate::Status::expired);
+}
+
+TEST_CASE("Test Certificate active status")
+{
+  // notBefore - 2 days older than 04/22/2022
```
Comment (bifurcation) [inline lib/hpke/test/certificate.cpp:518]:
```suggestion
  // notBefore - 2 days older than 04/22/2021
```

?

File: lib/hpke/test/certificate.cpp [hunk 4]
```diff
@@ -470,3 +472,64 @@ TEST_CASE("Test Subject Parsing")
...
+    "130831312d32322d3333302a300506032b65700321003beb77d095e9bb842d52f3642e4955"
+    "b907f1e42129be39db81059860e6ea5c27a34e304c300e0603551d0f0101ff0404030202a4"
+    "300f0603551d130101ff040530030101ff300d0603551d0e04060404b2e294f2301a060355"
+    "1d1104133011810f7573657240646f6d61696e2e636f6d300506032b6570034100ba82b764"
+    "43b126cab7db4a8a054e53ee25ef39d1ef2642931c7961c5591643edf73b25fc2bd7c3527b"
+    "e4e7d2b0606050b2e0edcfc8d6390b373e21f08116910b");
+
+  auto root = Certificate{ root_der };
+  auto status = root.status();
+  REQUIRE(status == Certificate::Status::expired);
```
Comment (bifurcation) [inline lib/hpke/test/certificate.cpp:513]:
```suggestion
  REQUIRE(root.status() == Certificate::Status::expired);
```

File: lib/hpke/test/certificate.cpp [hunk 5]
```diff
@@ -470,3 +472,64 @@ TEST_CASE("Test Subject Parsing")
...
+    "0405130831312d32322d3333302a300506032b65700321004b16ffec9d6db170f06ca73ec7"
+    "64e206c71c57c3f7454b4bd17bcef26c22b7f7a34e304c300e0603551d0f0101ff04040302"
+    "02a4300f0603551d130101ff040530030101ff300d0603551d0e0406040488fbad09301a06"
+    "03551d1104133011810f7573657240646f6d61696e2e636f6d300506032b6570034100c644"
+    "939be90b8121a7e5b371396f89b01235dd8cde01e88d6d7e09fb02ae6c1bb5e5abfe21de96"
+    "16a821444907e84cd4fb88167f1c3a4d4911f8260dafb21b05");
+
+  auto root = Certificate{ root_der };
+  auto status = root.status();
+  REQUIRE(status == Certificate::Status::active);
```
Comment (bifurcation) [inline lib/hpke/test/certificate.cpp:534]:
```suggestion
  REQUIRE(root.status() == Certificate::Status::active);
```

File: scripts/cert-gen.go [hunk 6]
```diff
@@ -17,16 +17,24 @@ import (
 )
 
 var (
+
```
Comment (bifurcation) [inline scripts/cert-gen.go:20]:
```suggestion
```

File: lib/hpke/include/hpke/certificate.h [hunk 7]
```diff
@@ -29,6 +29,14 @@ struct Certificate
 
   using ParsedName = std::map<int, std::string>;
 
+  // Certificate Status
+  enum struct Status
```
Comment (bifurcation) [inline lib/hpke/include/hpke/certificate.h:33]: Nit: I would call this `ExpirationStatus` and the methods `expiration_status`.

File: lib/hpke/include/hpke/certificate.h [hunk 8]
```diff
@@ -29,6 +29,14 @@ struct Certificate
...
+  {
+    expired,
+    inactive,
+    active
```
Comment (bifurcation) [inline lib/hpke/include/hpke/certificate.h:37]:
```suggestion
    inactive, // now < notBefore
    active,   // notBefore < now < notAfter
    expired,  // notAfter < now
```

**Review by bifurcation — APPROVED** [review]
Couple of cosmetic things, otherwise lgtm.  Thanks for putting this together @suhasHere

**Comment by bifurcation** [comment]
Argh.  @suhasHere I got too enthusiastic with the "Merge" button and merged without giving you a chance to fix my little cosmetic issues.  Could you please file a follow-on to fix those?

**Pull request merged**